### PR TITLE
Fix time loss detection and early retransmit

### PR DESCRIPTION
**Pull request by ianswett**

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -238,15 +233,20 @@ reordering_threshold:
...
+  
+time_reordering_fraction:
+: The reordering window as a fraction of max(smoothed_rtt, latest_rtt).
 
-use_time_loss:
-: When true, loss detection operates solely based on reordering
-  threshold in time, rather than in packet number gaps.
+loss_time:
+: The time at which the next packet will be considered lost based on early
+transmit or exceeding the reordering window in time.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: early transmit --> early retransmit

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -238,15 +233,20 @@ reordering_threshold:
...
 
 sent_packets:
-: An association of packet numbers to information about them, including a time
-  field indicating the time a packet was sent and a bytes field indicating the
-  packet's size.
+: An association of packet numbers to information about them, including a number
+  field indicating the packet number, a time field indicating the time a packet
+  was sent and a bytes field indicating the packet's size.  sent_packets is
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: oxford comma nit: "... was sent, and a bytes field ..."

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -238,15 +233,20 @@ reordering_threshold:
...
+  ordered by packet number, and packets remain in sent_packets until
+  acknowledged or lost.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: nit: "until they are acknowledged or lost"

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -258,8 +258,13 @@ follows:
...
    rto_count = 0
-   reordering_threshold = kReorderingThreshold
-   use_time_loss = false
+   if (UsingTimeLossDetection())
+     reordering_threshold = infinite
+     time_reordering_fraction = kTimeReorderingFraction
+   else:
+     reordering_threshold = kReorderingThreshold
+     time_reordering_fraction = infinite
+   lost_time = 0
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: loss_time = 0

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -406,10 +410,9 @@ Pseudocode for SetLossDetectionAlarm follows:
         alarm_duration = 2 * smoothed_rtt
       alarm_duration = max(alarm_duration, kMinTLPTimeout)
       alarm_duration = alarm_duration << handshake_count
-    else if (largest sent packet is acked):
-      // Early retransmit
-      // with an alarm to reduce spurious retransmits.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I think this comment can fit in a line

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -442,9 +445,9 @@ Pseudocode for OnLossDetectionAlarm follows:
        RetransmitAllHandshakePackets();
        handshake_count++;
      // TODO: Clarify early retransmit and time loss.
-     else if ():
+     else if (lost_time != 0):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: loss_time

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -463,9 +466,9 @@ Pseudocode for OnLossDetectionAlarm follows:
 ## Detecting Lost Packets
 
 Packets in QUIC are only considered lost once a larger packet number is
-acknowledged.  DetectLostPackets is called every time there is a new largest
-packet or if the loss detection alarm fires the previous largest acked packet is
-supplied.
+acknowledged.  DetectLostPackets is called every time an ack is received.
+If the loss detection alarm fires and the lost_time is set, the previous
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: Make loss_time or lost_time consistent. Either works. How about "reordering_threshold_in_time" and "reordering_threshold_in_packets" (or remove the "in" if you want more succinct) instead of loss_time and reordering_threshold ?

File: draft-ietf-quic-recovery.md [hunk 8]
```diff
@@ -463,9 +466,9 @@ Pseudocode for OnLossDetectionAlarm follows:
...
+largest acked packet is supplied.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This paragraph is a little discontinuous. Let's submit it and we can fix it up later

File: draft-ietf-quic-recovery.md [hunk 9]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
 Pseudocode for DetectLostPackets follows:
 
 ~~~
-   DetectLostPackets(acked):
+   DetectLostPackets(largest_acked):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: nit: "largest_acked" --> "largest_acked_number" ?

File: draft-ietf-quic-recovery.md [hunk 10]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
...
+     loss_time = 0
      lost_packets = {}
-     foreach (unacked less than acked):
-       time_delta = acked.time_sent - unacked.time_sent
-       packet_delta = acked.packet_number - unacked.packet_number
-       if (time_delta > kTimeReorderThreshold * smoothed_rtt):
+     time_when_lost = 0;
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This isn't the time but a wait period... suggestion: reordering_period ? or "short_timeout"? you could just call it short_timeout_period... up to you

File: draft-ietf-quic-recovery.md [hunk 11]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
...
+     if (time_reordering_fraction == infinite):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This should be !=

File: draft-ietf-quic-recovery.md [hunk 12]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
...
+       time_when_lost =
+         (1 + time_reordering_fraction) * max(latest_rtt, smoothed_rtt)
+     else if (largest_acked.packet_number == largest_sent_packet):
+       // Early retransmit alarm.
+       time_when_lost = 9/8 * max(latest_rtt, smoothed_rtt)
+     foreach (unacked less than largest_acked.packet_number):
+       time_since_sent = now() - unacked.time_sent
+       packet_delta = largest_acked.packet_number - unacked.packet_number
+       if (time_since_sent > time_when_lost):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: if time loss detection is disabled, won't time_when_lost be 0?

File: draft-ietf-quic-recovery.md [hunk 13]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
...
          lost_packets.insert(unacked)
        else if (packet_delta > reordering_threshold)
          lost_packets.insert(unacked)
+       else if (loss_time == 0 && time_when_lost != 0):
+         loss_time = time_when_lost - time_since_sent
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: add a comment here saying something like "loss_time is used for setting the loss detection alarm later."

File: draft-ietf-quic-recovery.md [hunk 14]
```diff
@@ -481,19 +484,31 @@ DetectLostPackets takes one parameter, acked, which is the largest acked packet.
...
      
-     // Inform the congestion controller of lost packets.
+     // Inform the congestion controller of lost packets and
+     // lets it decide whether to retransmit immediately.
      OnPacketsLost(lost_packets)
-     MaybeRetransmit(lost_packets)
+     foreach (packet in lost_packets)
+       sent_packets.remove(packet.packet_number)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: does it make sense, for editorializing, to move these packets to a "lost_packets" structure? doesn't need to be done here, but thinking aloud

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Done.

File: draft-ietf-quic-recovery.md (hunk 3 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: I left it as is, because it's shorter.

File: draft-ietf-quic-recovery.md (hunk 5 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: It was removed.

File: draft-ietf-quic-recovery.md (hunk 4 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: SGTM, I went with loss_time everywhere.

File: draft-ietf-quic-recovery.md (hunk 6 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Done

File: draft-ietf-quic-recovery.md (hunk 8 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: SGTM.

File: draft-ietf-quic-recovery.md (hunk 7 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: loss_time is somewhat orthogonal to reordering theshold.  Do you mean reordering_threshold and time_reordering_fraction, which are more analagous?

File: draft-ietf-quic-recovery.md (hunk 9 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: It's a packet, not a number, because I use fields of the packet below, such as the time it was sent.

File: draft-ietf-quic-recovery.md (hunk 11 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Thanks.

File: draft-ietf-quic-recovery.md (hunk 10 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: How about delay_until_lost?

File: draft-ietf-quic-recovery.md (hunk 12 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Good point, I set it to infinite.

File: draft-ietf-quic-recovery.md (hunk 14 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Possibly, though I still need a way to inform the congestion control of newly lost packets.  I was thinking of creating a pending_retransmissions data structure in congestion control to store any lost packets that couldn't be sent immediately?
